core/src/test/scala/org/broadinstitute/dsde/rawls/coordination/CoordinatedDataSourceActorSpec.scala

```diff
@@ -0,0 +1,196 @@
+    with ImplicitSender
+    with FlatSpecLike
+    with Matchers
+    with BeforeAndAfterAll
+    with ScalaFutures
+    with TableDrivenPropertyChecks
+    with MockitoSugar {
+  behavior of "CoordinatedDataSourceActor"
+
+  import system.dispatcher
+
+  override def afterAll {
+    TestKit.shutdownActorSystem(system)
+  }
+
+  private val tests = Table(
+    ("description", "function", "expected"),
+    (
+      "return a normal result",
+      () => 42,
+      Success(42),
+    ),
+    (
+      "not lose errors when they occur",
+      () => throw new RuntimeException("expected") with NoStackTrace,
+      Failure(new RuntimeException("expected")),
+    ),
+    (
+      "not wait for results that arrive too late",
+      () => {
+        Thread.sleep(10.seconds.toMillis)
+        "i'm running a bit late"
+      },
+      Failure(new TimeoutException("Futures timed out after [5 seconds]")),
+    ),
+  )
+
+  forAll(tests) { (description, function, expected) =>
+    it should description in {
+      val timeout = 5.seconds
+      val testActor = TestActorRef(CoordinatedDataSourceActor.props())
+      implicit val implicitAskTimeout: Timeout = Timeout(timeout * 2)
+      val mockSlickDataSource = mock[SlickDataSource]
+      val mockDataAccessFunction = mock[DataAccess => ReadWriteAction[Any]]
+      val transactionIsolation = TransactionIsolation.RepeatableRead
+      when(mockSlickDataSource.inTransaction(mockDataAccessFunction, transactionIsolation))
+        .thenReturn(Future(function()))
+      val future = testActor ?
+        CoordinatedDataSourceActor.Run(
+          mockSlickDataSource,
+          mockDataAccessFunction,
+          transactionIsolation,
+          Deadline.now + timeout,
+          timeout,
+        )
+      Await.ready(future, 30.seconds)
+      expected match {
+        case Success(expectedSuccess) =>
+          val Success(actualSuccess) = future.value.get
+          actualSuccess should be(expectedSuccess)
+        case Failure(expectedFailure) =>
+          val Failure(actualFailure) = future.value.get
+          expectedFailure.getClass.isAssignableFrom(actualFailure.getClass) should be(true)
+          actualFailure.getMessage should be(expectedFailure.getMessage)
+      }
+    }
+  }
+
+  it should "should run serially" in {
+    val individualSleep = 100.millis
+    val individualWaitTimeout = 1.second
+    val numTested = 100
+    implicit val implicitAskTimeout: Timeout = Timeout(individualWaitTimeout * numTested * 2)
+    val testActor = TestActorRef(CoordinatedDataSourceActor.props())
+    val mockSlickDataSource = mock[SlickDataSource]
+    val mockDataAccessFunction = mock[DataAccess => ReadWriteAction[Any]]
+    val transactionIsolation = TransactionIsolation.RepeatableRead
+
+    val quickening = new Quickening(individualSleep)
+    when(mockSlickDataSource.inTransaction(mockDataAccessFunction, transactionIsolation))
+      .thenAnswer(_ => Future(quickening.highlander()))
```

Comment:
What does `highlander()` do? I saw the code, but can't understand it's purpose

Reply:
The function throws an exception if there was more one `.highlander()` call at the same time.

[![There can be only one](https://user-images.githubusercontent.com/791985/80540525-82651e80-8977-11ea-9f06-906b3fcfe58e.png)](https://www.youtube.com/watch?v=sqcLjcSloXs)

When run by multiple threads, the function's internal `counter` will be greater than one if the callers were not synchronized correctly.